File: Ladder03/Java_code/InsertNodeInABinarySearchTree.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
// recursive version
public class Solution {
    /**
     * @param root: The root of the binary search tree.
     * @param node: insert this node into the binary search tree
     * @return: The root of the new binary search tree.
     */
    public TreeNode insertNode(TreeNode root, TreeNode node) {
        if (node == null) {
            return root;
        }
        if (root == null) {
            return node;
        }
        if (node.val > root.val) {
            root.right = insertNode(root.right, node);
        } else {
            root.left = insertNode(root.left, node);
        }
        return root;
    }
}

// non-recursive version
public class Solution {
    public TreeNode insertNode(TreeNode root, TreeNode node) {
        if (node == null) {
            return root;
        }
        if (root == null) {
            return node;
        }
        TreeNode p = root;
        while (p != null) {
            if (node.val > p.val) {
                if (p.right == null) {
                    p.right = node;
                    break;
                } else {
                    p = p.right;
                }
            } else {
                if (p.left == null) {
                    p.left = node;
                    break;
                } else {
                    p = p.left;
                }
            }
        }
        return root;
    }
}

File: NineChapters/03.BinaryTree_DivideConquer/Python_code/insert_node_in_a_binary_search_tree.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        this.val = val
        this.left, this.right = None, None
"""
# recursive version
class Solution:
    """
    @param root: The root of the binary search tree.
    @param node: insert this node into the binary search tree.
    @return: The root of the new binary search tree.
    """
    def insertNode(self, root, node):
        if root is None:
            return node
        if node is None:
            return root
        if node.val > root.val:
            root.right = self.insertNode(root.right, node)
        elif node.val < root.val:
            root.left = self.insertNode(root.left, node)
        return root

# non-recursive version
class Solution:
    def insertNode(self, root, node):
        if root is None:
            return node
        if node is None:
            return root
        p = root
        while p:
            if node.val > p.val:
                if p.right is None:
                    p.right = node
                    return root
                else:
                    p = p.right
            else:
                if p.left is None:
                    p.left = node
                    return root
                else:
                    p = p.left
        return root